Enable reindexing of indices from packages in Storage Service

tw4l opened the pull request:
This PR introduces a new management task `rebuild_aip_index_from_storage_service.py` for reindexing AIPs from packages in the Storage Service, and adds a `--reindex-from-storage-service` flag to `rebuild_transfer_backlog.py` which enables the same for Transfers. Collectively, these changes allow for reindexing from any non-deleted Transfers and AIPs/AICs in the Storage Service, including those stored in encrypted storage locations.

Both tasks poll the Storage Service API to determine which packages should be indexed, filtering out deleted packages and those not associated with the specified origin pipeline (defaulting to the current pipeline if a value is not provided).

`rebuild_aip_index_from_storage_service.py` downloads a copy of the AIP METS file (and the AIC METS file, for AICs) into a temporary directory and uses that locally downloaded METS file for reindexing. I've made this its own task as there were substantial enough changes with the existing task that it seemed to warrant being separated, but I would be happy to merge them together if that is preferred. 

`rebuild_transfer_backlog.py`, when called with the  `--reindex-from-storage-service` flag, downloads a copy of each transfer package from the Storage Service into a temporary directory, uses that locally downloaded copy to index the transfer, and then immediately deletes the local copy when it is no longer needed.

These tasks are slower than the existing indexing tasks and could be made more efficient. The reasons for the performance hit seem to be that:
* Although `rebuild_aip_index_from_storage_service.py` downloads the METS file and not the whole AIP, the Storage Service decrypts the entire package by moving a copy of the package to an internal Storage Service location in order to support downloading this one file.
* It would likely be more efficient (assuming more performant decryption for downloading individual files, if the issue described immediately above is fixed) to download the transfer METS file and use that as the source of information for indexing rather than downloading the entire transfer, like how AIPs are handled. This would require modifying the indexing functions for transfers in `elasticSearchFunctions.py` to support reading from a transfer METS file for information about files with a transfer rather than walking the transfer directory.

This PR also introduces a few helper functions in `archivematicaFunctions.py`, `fileOperations.py`, and `storageService.py` to support the new changes,

Fixes https://github.com/archivematica/issues/issues/734 (however, some problems also noted in that issue, such as not being able to arrange and create a SIP from encrypted transfers in the Arrangement tab, are still outstanding as they are not directly related to reindexing and should be addressed in separate issues. I will create those issues as needed as a follow-up to creating this PR and link them below).

Related to https://github.com/archivematica/Issues/issues/1228

tw4l committed:
Enable rebuilding indexes from Storage Service

This commit enables reindexing of transfers and AIPs/AICs via the
Storage Service API. This enables reindexing of packages stored in
encrypted locations as well as some remote locations.

Calling `rebuild_transfer_backlog` with the `--from-storage-service`
argument will reindex non-deleted transfers associated with a specific
origin pipeline by temporarily downloading a copy of each transfer for
indexing.

The new `rebuild_aip_index_from_storage_service` task will reindex non-
deleted AIPs and AICs associated with a specific origin pipeline by
temporarily downloading and indexing from copies of the AIP and AIC
METS files.

tw4l commented:
Related recently-created issues:
* https://github.com/archivematica/Issues/issues/1244
* https://github.com/archivematica/Issues/issues/1245
* https://github.com/archivematica/Issues/issues/1246
* https://github.com/archivematica/Issues/issues/1247

replaceafill commented on src/archivematicaCommon/lib/archivematicaFunctions.py:
```diff
@@ -434,3 +455,61 @@ def find_transfer_path_from_ingest(transfer_path, shared_path):
+        chars_to_remove += len(suffix)
+    # Check if we have characters to remove to avoid accidentally returning an
+    # empty string with name[:-0].
+    if not chars_to_remove > 0:
```
```suggestion
    if not chars_to_remove:
```

replaceafill commented on src/archivematicaCommon/lib/archivematicaFunctions.py:
```diff
@@ -434,3 +455,61 @@ def find_transfer_path_from_ingest(transfer_path, shared_path):
+    return mets_path
+
+
+def filter_packages(package_list, pipeline_uuid=None):
```
The call to this function from `rebuild_transfer_backlog.py` which doesn't have a comment explaining that this will filter out packages with `DELETED` status is a bit unclear. Maybe it could be renamed to `filter_deleted_packages` or pass an iterable explicitly with the status you want to filter out?

tw4l replied:
I refactored the function a bit and renamed it to `filter_packages_by_status_and_pipeline`, which I think makes it clearer.

replaceafill commented on src/archivematicaCommon/tests/test_archivematica_functions.py:
```diff
@@ -40,10 +40,42 @@ def test_get_bag_size_bag_missing_oxum(tmpdir):
+        assert package_name == test_package["package_name"]
+
+        package_name_without_uuid = am.package_name_from_path(
+            current_path, remove_uuid_suffix=True
```
Is it worth adding a test case with `remove_uuid_suffix=False`?

tw4l replied:
Agreed and done.

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py:
```diff
@@ -166,22 +188,70 @@ def populate_data(self, es_client, transfer_backlog_dir):
+                )
+                continue
+            # Transfers are downloaded as .tar files, so we extract files before indexing.
+            extract_package(local_package, temp_backlog_dir)
```
Should potential errors in the extraction be handled here?

replaceafill replied:
This stills assumes extraction worked, correct?

I was wondering what would happen if the `subprocess.call` fails with a return code other than `0`. Should we at least log that?

tw4l replied:
Yes, good point! I had taken a more indirect route of throwing an error when the directory expected to be extracted wasn't present, but you're right that it would be better to explicitly check the return code. I will add that before merging.

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py:
```diff
@@ -166,22 +188,70 @@ def populate_data(self, es_client, transfer_backlog_dir):
+                continue
+            # Transfers are downloaded as .tar files, so we extract files before indexing.
+            extract_package(local_package, temp_backlog_dir)
+            local_package_sans_extension = am.package_name_from_path(local_package)
```
Do you think `_without_` might be clearer than `_sans_`?

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py:
```diff
@@ -0,0 +1,326 @@
+            es.setup_reading_from_conf(django_settings)
+            es_client = es.get_client()
+        except ElasticsearchException:
+            self.error("Error: Elasticsearch may not be running")
```
Is it worth logging the exception like `rebuild_transfer_backlog.py` does?

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py:
```diff
@@ -0,0 +1,326 @@
+        )
+
+        if delete_before_reindexing is True:
+            es.delete_aip(es_client, uuid)
```
Is it worth logging which package is about to be deleted like `rebuild_elasticsearch_aip_index_from_files.py` does?

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py:
```diff
@@ -70,11 +77,17 @@ def add_arguments(self, parser):
             "--transfer-backlog-dir", default=self.DEFAULT_TRANSFER_BACKLOG_DIR
         )
         parser.add_argument("--no-prompt", action="store_true")
+        parser.add_argument("--reindex-from-storage-service", action="store_true")
```
Not sure if it's already planned in the scope of this project, but maybe it's worth filing a documentation issue about updating the `Rebuild the indexes` section of the docs to include these new arguments.

tw4l replied:
Yes, good point!

replaceafill replied:
I noticed you filed https://github.com/archivematica/Issues/issues/1254 Thanks!

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py:
```diff
@@ -166,22 +188,70 @@ def populate_data(self, es_client, transfer_backlog_dir):
         self.success("{} transfers indexed!".format(processed))
 
+    def populate_data_from_storage_service(self, es_client, pipeline_uuid):
+        """Populate data in search indices and/or database from files temporarily downloaded from Storage Service.
```
This might require further discussion but I think the introduction of `black` has changed what we consider long lines. Docstrings and comments are also a bit special. and [pep8 has](https://www.python.org/dev/peps/pep-0008/#id19):

> ... it is okay to increase the line length limit up to 99 characters, provided that comments and docstrings are still wrapped at 72 characters.

tw4l replied:
Good catch -  I didn't realize that the 72 character limit was strict for docstrings and comments. I've reworked this docstring and the others so that they wrap at 72 characters.

replaceafill reviewed (commented):
@tw4l I think this is on the right track and I really like the METS approach. I think most of my comments don't require significant changes.

However, although I agree with you that the new functionality is substantial enough to have its own command, I think the duplication of code between the new task and `rebuild_elasticsearch_aip_index_from_files` will make maintenance harder.

The `get_aips_in_aic` functions for example have some common functionality. Or the parsing of common arguments (`--delete` and `--delete-all`), and the initial part of the `handle` method (where Elasticsearch is tested and the `--delete-all` argument is handled). Do you think it's worth extracting some of these? Or maybe file an issue and do it later?

tw4l commented:
Hi @replaceafill - thank you for taking a look at this, and for the helpful suggestions. I think I've addressed your comments in the follow-up commits, including pulling out some of the task setup and shared `get_aips_in_aic` logic into shared helpers as requested to help with maintenance. I also added code in `rebuild_aip_index_from_storage_service.py` to delete METS files immediately after indexing to avoid any potential disk space issues, and renamed the `--reindex-from-storage-service` flag to just `--from-storage-service` to make it a little less unwieldy. Please let me know if you'd like to see any further changes.

replaceafill commented on src/archivematicaCommon/lib/archivematicaFunctions.py:
```diff
@@ -496,11 +536,15 @@ def relative_path_to_aip_mets_file(uuid, current_path):
-def filter_packages(package_list, pipeline_uuid=None):
-    """Filter out deleted packages and packages from other pipelines.
+def filter_packages_by_status_and_pipeline(
+    package_list, valid_statuses=["UPLOADED", "DEL_REQ"], pipeline_uuid=None
```
Even though the function doesn't mutate `valid_statuses` I think it's worth to use a `tuple` for default arguments. Alternatively you could use `None` and then set its initial value. See https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

tw4l replied:
TIL!

replaceafill commented on src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py:
```diff
@@ -248,8 +257,16 @@ def populate_data_from_storage_service(self, es_client, pipeline_uuid):
+                    transfer_indexed = True
             shutil.rmtree(temp_backlog_dir)
-            processed += 1
+            if transfer_indexed is True:
```
```suggestion
            if transfer_indexed:
```

replaceafill reviewed (approved):
Thank you @tw4l! LGTM :+1:
